### PR TITLE
[AutoDiff] Robust mangling support for AD associated functions.

### PR DESCRIPTION
**Pull request by dan-zheng**
`Demangler`/`Remangler` support for AD associated functions:
JVPs, VJPs, differentials, pullbacks.

These functions now appear with clarifying annotations in SIL:
```
// VJP wrt 0 source 0 for foo(_:)
sil hidden @$s8mangling3fooyS2fFTZp0r0
```

The following components are mangled:
- Original function name.
- Parameter indices.
- Result index.

Resolves [TF-679](https://bugs.swift.org/projects/TF/issues/TF-679).

### CODE REVIEW

**Review by bartchr808 — APPROVED** [review]

File: lib/Demangling/Remangler.cpp [hunk 1]
```diff
@@ -1954,6 +1957,47 @@ void Remangler::mangleReabstractionThunkHelperWithSelf(Node *node) {
...
+  Buffer << 'r' << node->getIndex();
+}
+
+void
+Remangler::mangleAutoDiffAssociatedFunctionHelper(Node *node, StringRef op) {
+  assert(node->getNumChildren() == 3);
+  mangleChildNode(node, 0); // original function
+  Buffer << op;
+  mangleChildNode(node, 1); // wrt parameter indices
+  mangleChildNode(node, 2); // result index
```
Comment (rxwei) [inline lib/Demangling/Remangler.cpp]: Add a TODO for mangling generic signature?

**Review by rxwei — APPROVED** [review]
This is one of the best things that have ever happened to AD :)

File: lib/Demangling/Remangler.cpp (hunk 1 above)
Comment (dan-zheng) [reply lib/Demangling/Remangler.cpp]: Done in https://github.com/apple/swift/pull/26624/commits/e8a3e55f190cad3d0344f164ef5e440816015173.

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

File: lib/AST/ASTMangler.cpp [hunk 2]
```diff
@@ -369,51 +369,86 @@ std::string ASTMangler::mangleReabstractionThunkHelper(
   return finalize();
 }
 
+// SWIFT_ENABLE_TENSORFLOW
+/// Get a `NodePointer` representing an autodiff associated function for the
+/// given original function mangled name, associated function node kind,
+/// and parameter indices.
+static NodePointer getMangledAutoDiffAssociatedFunctionNode(
```
Comment (dan-zheng) [inline lib/AST/ASTMangler.cpp]: Note for the curious: this logic was adapted from [`SpecializationMangler::finalize`](https://github.com/apple/swift/blob/4df23c8e12f718b47729d5546a0fe188eab1d490/lib/SILOptimizer/Utils/SpecializationMangler.cpp#L49).

**Pull request merged**